### PR TITLE
Optimize to get more speed from Oj

### PR DESCRIPTION
**Pull request by ohler55**

### CODE REVIEW

**Comment by ohler55** [comment]
This give a pretty decent speed boost.

File: benchmark.rb [hunk 1]
```diff
@@ -22,13 +21,15 @@
 obj.delete(:'ActiveRecord')
 obj.delete(:Complex)
 
+Oj::Rails.optimize(Array, BigDecimal, Hash, Range, Regexp, Time, ActiveSupport::TimeWithZone, User)
```
Comment (stereobooster) [inline benchmark.rb:24]: Nobody will use `Oj::Rails.optimize(User)`

Comment (ohler55) [reply benchmark.rb:24]: Their choice I suppose. It does improve performance.

Comment (stereobooster) [reply benchmark.rb:24]: It is too hard. Users install `oj` and hope everything is done. It took me some time to realise and prove that `to_json` inside Rails project is slow. It is not a choice. It is simple lack of knowledge.

**Comment by ohler55** [comment]
ok, so do you want to see changes in the PR?

**Comment by ohler55** [comment]
As an alternative the optimized versions could be the default that would have to be turned off if the user wanted to modify the default as_json method.

File: benchmark.rb (hunk 1 above)
Comment (ohler55) [reply benchmark.rb:24]: Do you want me to remove User?

**Comment by stereobooster** [comment]
I added Oj::Rails.encode [test_report](https://github.com/stereobooster/ruby-json-benchmark/blob/master/test_report.txt).

Not sure why we need `Oj::Rails.encode`. Why another entry-point, when there is already `Oj.dump`. This is potential source of errors and confusion.

About `Oj::Rails.optimize(Array, BigDecimal, Hash, Range, Regexp, Time, ActiveSupport::TimeWithZone)` I expect that Oj will provide `optimize_rails`

```ruby
def optimize_rails
  Oj::Rails.set_encoder()
  Oj::Rails.set_decoder()
  Oj::Rails.optimize(Array, BigDecimal, Hash, Range, Regexp, Time, ActiveSupport::TimeWithZone)
end
```

**Comment by ohler55** [comment]
Odd that Complex and Rational are not matching. I'll double check to make sure that if json/adds are called that the Oj add_jsons are also called.

Oj::Rails.encode does the same as dump with the mode option set to rails but cuts out a few stack calls. It doesn't imply to the user that the other dump options can still be used.

Not being a Rails expert I figure most people would actually set the encoder and decoder and then forget about it so encoding would be ActiveSupport::JSON.encode(). Since that could not be used for the benchmark without forcing Oj for the other runs Oj::Rails.encode() seemed like the closest call.

I understand what you were asking for before with the optimize_rails. I thought you just wanted to rename the Oj::Rails.optimize. I will add the Oj.optimize_rails() on the next Oj release that calls those three functions.

**Comment by ohler55** [comment]
I think you might have resolve the conflicts since I don't have write access. That or just drop the PR and make what ever changes you deem appropriate.

**Comment by ohler55** [comment]
Please add ```Oj.add_to_json(Rational, Complex)``` to the tests so that Oj matches the 
```
require 'json/add/complex'
require 'json/add/rational'
```
that the json gem uses.

I like the results though. A 7x performance boost over the regular Rails.

**Comment by stereobooster** [comment]
I added main bits of this PR to master, except (`User`), so I will close this.

As of optimisation of ActiveRecord objects I'm in doubts. On one side there is 

```ruby
ObjectSpace.each_object(ActiveRecord::Base).select { |klass| klass < self }
```

on the other side we do not want to use a lot of magic

**Comment by stereobooster** [comment]
So it is `Oj.add_to_json(Rational, Complex)`. I confused and put `Oj::Rails.optimize(Array, BigDecimal, Hash, Range, Regexp, Time, Complex, Rational)`.

**Comment by ohler55** [comment]
I'm not sure of the ActiveRecord::Base objects. It was an easy addition as default is to encode the @attributes. Lets see if it gets used.

You need th eOj::Rails.optimize() for the rails stuff. If you also require the json/add/xxx then the Oj.add_to_json should be called to match those requires.

**Comment by stereobooster** [comment]
Updated tests and benchmarks. Added comparison of speed by ips

**Comment by ohler55** [comment]
Super. Looks great. Thanks.

**Pull request closed without merging**